src/Sulu/Component/CustomUrl/Manager/CustomUrlManager.php

```diff
@@ -0,0 +1,158 @@
+{
+    /**
+     * @var DocumentManagerInterface
+     */
+    private $documentManager;
+
+    /**
+     * @var CustomUrlRepository
+     */
+    private $customUrlRepository;
+
+    /**
+     * @var PathBuilder
+     */
+    private $pathBuilder;
+
+    public function __construct(
+        DocumentManagerInterface $documentManager,
+        CustomUrlRepository $customUrlRepository,
+        PathBuilder $pathBuilder
+    ) {
+        $this->documentManager = $documentManager;
+        $this->customUrlRepository = $customUrlRepository;
+        $this->pathBuilder = $pathBuilder;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function create($webspaceKey, array $data, $locale = null)
+    {
+        $document = $this->documentManager->create('custom_urls');
+        $this->bind($document, $data, $locale);
+
+        $this->documentManager->persist(
+            $document,
+            $locale,
+            [
+                'parent_path' => $this->getItemsPath($webspaceKey),
+                'node_name' => Urlizer::urlize($document->getTitle()),
+                'load_ghost_content' => true,
+            ]
+        );
+
+        return $document;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function readList($webspaceKey)
+    {
+        // TODO pagination
+
+        return $this->customUrlRepository->findList($this->getItemsPath($webspaceKey));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function read($uuid, $locale = null)
+    {
+        return $this->documentManager->find($uuid, $locale, ['load_ghost_content' => true]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function update($uuid, array $data, $locale = null)
+    {
+        $document = $this->read($uuid, $locale);
+        $this->bind($document, $data, $locale);
+
+        $this->documentManager->persist($document, $locale);
+
+        return $document;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function delete($uuid)
+    {
+        $this->documentManager->remove($this->read($uuid));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getFields()
+    {
+        return [
+            'title' => ['property' => 'title'],
+            'published' => ['property' => 'published'],
+            'baseDomain' => ['property' => 'baseDomain'],
+            'domainParts' => ['property' => 'domainParts', 'type' => 'json_array'],
+            'target' => ['property' => 'target', 'type' => 'reference'],
+            'multilingual' => ['property' => 'multilingual'],
+            'canonical' => ['property' => 'canonical'],
+            'redirect' => ['property' => 'redirect'],
+            'targetLocale' => ['property' => 'targetLocale'],
+        ];
+    }
+
+    /**
+     * Bind data array to given document.
+     *
+     * @param CustomUrlDocument $document
+     * @param array $data
+     * @param string $locale
+     */
+    private function bind(CustomUrlDocument $document, $data, $locale)
```

Comment:
Agreed. The mapper component is already existing in Sulu, its called the
FormFactory: https://github.com/sulu-io/sulu/blob/develop/src/Sulu/Component/Content/Mapper/ContentMapper.php#L231

I think you could actually simply use that.

On Fri, Jan 22, 2016 at 02:02:30AM -0800, Wachter Johannes wrote:
>    In [1]src/Sulu/Component/CustomUrl/Manager/CustomUrlManager.php:
> 
>  > +            'target' => ['property' => 'target', 'type' => 'reference'],
>  > +            'multilingual' => ['property' => 'multilingual'],
>  > +            'canonical' => ['property' => 'canonical'],
>  > +            'redirect' => ['property' => 'redirect'],
>  > +            'targetLocale' => ['property' => 'targetLocale'],
>  > +        ];
>  > +    }
>  > +
>  > +    /**
>  > +     * Bind data array to given document.
>  > +     *
>  > +     * @param CustomUrlDocument $document
>  > +     * @param array $data
>  > +     * @param string $locale
>  > +     */
>  > +    private function bind(CustomUrlDocument $document, $data, $locale)
> 
>    the code i have written here could also be extracted in an own class when
>    we add a few exception handling and the validation component of symfony
>    here we could acchieve a "generic" data to document mapper. this would be
>    in my opinion the best solution who calls this mapping is not important i
>    think.
> 
>    The best solution for me it would be to create a mapper component (or find
>    one) which will be feeded with the document metadata that maps arrays of
>    data to a document. this will be called by the controller or even a
>    command. the returned document will be passed to a manager which interacts
>    with the document manager to encapsulate this.
> 
>    —
>    Reply to this email directly or [2]view it on GitHub.
> 
>    Reverse link: [3]unknown
> 
> References
> 
>    Visible links
>    1. https://github.com/sulu-io/sulu/pull/1927#discussion_r50519640
>    2. https://github.com/sulu-io/sulu/pull/1927/files#r50519640
>    3. https://github.com/sulu-io/sulu/pull/1927/files#r50519640